### PR TITLE
Ruby: port `py/weak-sensitive-data-hashing`

### PR DESCRIPTION
**Pull request by alexrford**
This is a pretty direct port of `py/weak-sensitive-data-hashing`. The main difference is that the sources are instances of `SensitiveNode` rather than the python `SensitiveDataSource`. These are similar concepts which partially share an interface, but the python version may be a bit more sophisticated than what we currently have for ruby.

### CODE REVIEW

File: ruby/ql/lib/codeql/ruby/security/WeakSensitiveDataHashingQuery.qll [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+ * computationally expensive hash function.
+ */
+module NormalHashFunction {
+  import WeakSensitiveDataHashingCustomizations::NormalHashFunction
+
+  /**
+   * A taint-tracking configuration for detecting use of a broken or weak
+   * cryptographic hashing algorithm on sensitive data.
+   */
+  class Configuration extends TaintTracking::Configuration {
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/security/WeakSensitiveDataHashingQuery.qll:29]: May as well use the new API for this.

File: ruby/ql/lib/codeql/ruby/security/WeakSensitiveDataHashingQuery.qll [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  import WeakSensitiveDataHashingCustomizations::ComputationallyExpensiveHashFunction
+
+  /**
+   * A taint-tracking configuration for detecting use of a broken or weak
+   * cryptographic hashing algorithm on passwords.
+   *
+   * Passwords has stricter requirements on the hashing algorithm used (must be
+   * computationally expensive to prevent brute-force attacks).
+   */
+  class Configuration extends TaintTracking::Configuration {
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/security/WeakSensitiveDataHashingQuery.qll:63]: Same.

**Review by hvitved — CHANGES_REQUESTED** [review]

File: ruby/ql/lib/codeql/ruby/security/WeakSensitiveDataHashingQuery.qll [hunk 3]
```diff
@@ -60,19 +56,28 @@ module ComputationallyExpensiveHashFunction {
...
+ * Provides a path graph for tracking use of a broken or weak cryptographic
+ * hashing algorithm on sensitive data.
+ */
+module WeakSensitiveDataHashing {
+  /**
+   * A path graph for tracking use of a broken or weak cryptographic
+   * hashing algorithm on sensitive data.
+   */
+  module Config =
+    DataFlow::MergePathGraph<NormalHashFunction::PathNode, ComputationallyExpensiveHashFunction::PathNode, NormalHashFunction::PathGraph, ComputationallyExpensiveHashFunction::PathGraph>;
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/security/WeakSensitiveDataHashingQuery.qll:82]: Nice!

**Comment by alexrford** [comment]
Performance is a bit borderline - looking into removing IDs as sources both to reduce FPs and potentially improve performance.

**Comment by alexrford** [comment]
Replaced by https://github.com/github/codeql/pull/16781

**Pull request closed without merging**